include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,203 @@
+	const float cosFactor = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
 
-	//speculative execution
-	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
-	params_t params = instr_getParameters(instr, bsdf_data);
-	float bxdf_eval_scalar_part;
-	uint ndf = instr_getNDF(instr);
-	float a = params_getAlpha(params);
-	float a2 = a*a;
-#ifndef ALL_ISOTROPIC_BXDFS
-	float ay = params_getAlphaV(params);
-	float ay2 = ay*ay;
-#endif
+	uvec3 regs = instr_decodeRegisters(instr);
+	mat2x3 ior;
+	mat2x3 ior2;
+	params_t params;
+	irr_glsl_AnisotropicMicrofacetCache microfacet;
+	bsdf_data_t bsdf_data;
 
-	float cosFactor = op_isBSDF(op) ? abs(currBSDFParams.isotropic.NdotL):max(currBSDFParams.isotropic.NdotL,0.0);
+	bxdf_eval_t bxdf_eval = bxdf_eval_t(0.0);
 
-	if (cosFactor>FLT_MIN && op_hasSpecular(op))
+	if (cosFactor > FLT_MIN)
 	{
+		//speculative execution
+		bsdf_data = fetchBSDFDataForInstr(instr);
+		ior = bsdf_data_decodeIoR(bsdf_data, op);
+		ior2 = matrixCompMult(ior, ior);
+		params = instr_getParameters(instr, bsdf_data);
+		const float ior_scalar = dot(CIE_XYZ_Luma_Y_coeffs, ior[0]);
+		float bxdf_eval_scalar_part;
+		uint ndf = instr_getNDF(instr);
+		float a = params_getAlpha(params);
+		float a2 = a*a;
+		float ay = params_getAlphaV(params);
+		float ay2 = ay*ay;
+		const vec3 albedo = params_getReflectance(params);
+
+		bool is_valid = op_isBXDF(op);
+		bool refraction = false;
+#ifndef NO_BSDF
+		//here actually using stronger check for BSDF because it's probably worth it
+		if (op_isBSDF(op) && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+		{
+			is_valid = irr_glsl_calcAnisotropicMicrofacetCache(microfacet, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, ior_scalar, 1.0/ior_scalar);
+			refraction = true;
+		}
+		else
+#endif
+		{
+			microfacet = _microfacet;
+		}
+
+		if (is_valid)
+		{
+#if defined(OP_DIFFUSE) || defined(OP_DIFFTRANS)
+			if (op_isDiffuse(op))
+			{
+				bxdf_eval = albedo * irr_glsl_oren_nayar_cos_eval(s, currInteraction.isotropic, a2);
+			} else
+#endif
+#if defined(OP_CONDUCTOR) || defined(OP_DIELECTRIC)
+			if (op_hasSpecular(op))
+			{
+			BEGIN_CASES(ndf)
 #ifdef NDF_GGX
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_GGX) {
-#endif
-
+				CASE_BEGIN(ndf, NDF_GGX) {
 #ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
+					bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(s, microfacet.isotropic, currInteraction.isotropic, a2);
 #else
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(currBSDFParams, currInteraction, a, ay);
-#endif
-
-
-#ifndef ONLY_ONE_NDF
-		} else
+					bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(s, microfacet, currInteraction, a, a2, ay, ay2);
```

Comment:
dont use `cos_eval_DG` .. break them down just like you did the remainder functions in the remainder stream